### PR TITLE
Fix for overly eager package object search.

### PR DESCRIPTION
**Pull request by paulp**
A subtle change in the order in which symbol attributes were
inspected (now you know why I avoid vals in the compiler) led to a
cycle during initialization for slick. I'm afraid I don't know how
to reproduce the issue outside of slick and sbt, so I added some
logging instead.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Contexts.scala [hunk 1]
```diff
@@ -761,26 +761,36 @@ trait Contexts { self: Analyzer =>
...
+            uninitialized("" + module)
         }
+        else uninitialized("" + pkg)
       )
       def inPackageObject(sym: Symbol) = (
            !sym.isPackage
-        && !sym.owner.isPackageClass
         && (sym.owner ne NoSymbol)
-        && (sym.owner.owner == pkgClass || matchesInfo)
+        && sym.owner.isPackageObjectClass
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: this (correct) change is not just about initialization order, right?

in this light, the commit message could be improved by pointing out this additional fix

Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: No, it is just about initialization order. I realize that is not entirely obvious.

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: I was talking about switching from `sym.owner.isPackageClass` to `sym.owner.isPackage*Object*Class`

I can't imagine how that would affect initialization. I tried closing my eyes and visualizing it with my mind's eye. Thinking out of the box. Being mindful. Nope. Could you make it illuminated?

Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: I know.  I meant that the only changes with any meaning in this commit are initialization order. It actually changes from !sym.owner.isPackageClass to sym.owner.isPackageObjectClass (you omitted an important negation) and in this context, those are equivalent. "This context" means "when the only question is whether sym is in the package object of package pkg." If sym.owner.isPackageClass, it cannot be; if sym.owner.isPackageObjectClass it must be, subject to the remaining condition that sym.owner.owner == pkg(Class).

Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: Of course I have to retract some of that. The funny thing is that this represents me unfixing a bug; the use of isPackageObjectClass fell out of my restoring the old implementation to see if it fixed the cycle (it did) but then I also restored some of its bugs.

Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: The retraction, for completeness, is that the above is only true if sym is not inherited. My earlier change to !isPackageClass is for exactly this reason; the inherited symbol has to get a shot at matchesInfo.

Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: Which leaves me wondering why the tests passed.

**Comment by paulp** [comment]
PLS REBUILD ALL

**Comment by paulp** [comment]
Is there an award for taking a pull request which solves the problem and for which all tests pass, imagining a problem, and then updating it so it breaks horribly?

**Comment by paulp** [comment]
Going to reopen.

**Pull request closed without merging**